src/compiler/checker.ts

```diff
@@ -5765,12 +5765,13 @@ namespace ts {
             return type.modifiersType;
         }
 
+        function isPartialMappedType(type: Type) {
+            return getObjectFlags(type) & ObjectFlags.Mapped && !!(<MappedType>type).declaration.questionToken;
+        }

```

Comment:
This seems like an awfully specific check - does it work on homomorphic mapped types in which all properties of the type being mapped are already optional? For example:

```ts
type Shapify<T> = {
    [K in keyof T]: any;
}

interface MyInterface {
  something?: number;
}

class MyClass<T extends MyInterface> {
  doIt(data: Shapify<T>) {}
}

class MySubClass extends MyClass<MyInterface> {}

function fn(arg: typeof MyClass) {};

fn(MySubClass);
```

The same question applies to if you used `Readonly<T>` or others.